lxd/container_lxc.go

```diff
@@ -2711,6 +2711,105 @@ func (c *containerLXC) setupPhysicalParent(deviceName string, m types.Device) (s
 	return hostName, nil
 }
 
+// detachInterfaceRename enters the container's network namespace and moves the named interface
+// in ifName back to the network namespace of the running process as the name specified in hostName.
+func (c *containerLXC) detachInterfaceRename(netns string, ifName string, hostName string) error {
+	pid := os.Getpid()
+
+	_, err := shared.RunCommand("nsenter", fmt.Sprintf("--net=%s", netns), "ip", "link", "set", ifName, "name", hostName, "netns", fmt.Sprintf("%d", pid))
```

Comment:
I don't like that we entirely rely on the interface name in the container to have been left intact by the user.

Say you have container that's setup with:
 - eth0 => bridged
 - eth1 => physical

As root in the container I could swap their names, causing LXD to move the veth to the host and then the kernel to move the physical interface to the host.

This now effectively gives me a configured network interface on the host which is attached to a bridge that I can have other containers on, if properly configured, I could use that to take over some specific host traffic.

In general LXD should assume a hostile user in the container and use ways of tracking resources which cannot be modified by the user.

I also think that we should ensure we do an IPv4/IPv6 reset on the interface as it's removed from the container, or at least ensure it's brought down as part of this.

Reply:
@stgraber as discussed in IRC, I will modify this to use forknet and to ensure the nic is down before being removed, and also having its IPs removed.